admin_ui/app/permission/PermissionEditPage.js

```diff
@@ -1,74 +1,96 @@
-import React from 'react';
+import React, { useCallback, useEffect } from 'react';
 import PropTypes from 'prop-types';
-import { Link } from '@reach/router';
+import { navigate } from '@reach/router';
 import useDocumentTitle from '@rehooks/document-title';
-import Input from '../../components/Input';
-import Textarea from '../../components/Textarea';
-import Button from '../../components/Button';
-import Select from 'react-select';
+import { useDispatch, useSelector } from 'react-redux';
+import find from 'lodash/find';
+import PermissionDeleteModal from '../modals/PermissionDeleteModal';
+import { openPermissionDeleteModal } from '../modals/permissionModalsStore';
+import PermissionForm from './PermissionForm';
+import { updatePermission, getPermissionInfo, setPermissionFormValues } from './permissionStore';
+import LoadingIndicator from '../../components/LoadingIndicator';
+import yeepClient from '../yeepClient';
+
+function gotoPermissionList() {
+  navigate('/permissions');
+}
 
 const PermissionEditPage = ({ permissionId }) => {
+  const records = useSelector((state) => state.permission.list.records);
+  const isLoading = useSelector((state) => state.permission.form.isLoading);
+  const dispatch = useDispatch();
+
   useDocumentTitle(`Edit permission #${permissionId}`);
+
+  useEffect(() => {
+    // check if permission already exists in store
+    const permission = find(records, (e) => e.id === permissionId);
+
+    if (permission) {
+      dispatch(setPermissionFormValues(permission));
+    } else {
+      // role does not exist in memory - retrieve from API
+      dispatch(getPermissionInfo({ id: permissionId })).then((permission) => {
+        dispatch(setPermissionFormValues(permission));
+      });
+    }
+
+    return () => {
+      yeepClient.redeemCancelToken(getPermissionInfo);
+    };
+  }, [permissionId, records, dispatch]);
+
+  const onPermissionDelete = useCallback(
+    (values) => {
+      dispatch(
+        openPermissionDeleteModal({
+          role: {
+            id: permissionId,
+            ...values,
+          },
+        })
+      );
+    },
+    [dispatch, permissionId]
+  );
+
+  const submitForm = useCallback(
+    (values) => {
+      dispatch(
+        updatePermission({
+          id: permissionId,
+          name: values.name,
+          description: values.description,
+        })
+      ).then((isPermissionUpdated) => {
+        if (isPermissionUpdated) {
```

Comment:
Hm, there is no error action here. In case of an error, the related `errors` object will be populated and display error messages to the user on screen.